hpx/parallel/spmd_block.hpp

```diff
@@ -0,0 +1,195 @@
+        {
+           return barrier_.get().wait(hpx::launch::async);
+        }
+
+    private:
+        std::size_t num_images_;
+        std::size_t image_id_;
+        mutable std::reference_wrapper<hpx::lcos::barrier> barrier_;
+    };
+
+    namespace detail
+    {
+        template <typename F>
+        struct spmd_block_helper
+        {
+            std::string name_;
+            std::size_t num_images_;
+
+            template <typename ... Ts>
+            void operator()(std::size_t image_id, Ts && ... ts) const
+            {
+                printf("image_id is equal to %lu and num_images is %lu\n", image_id, num_images_);
+                std::cout << "The barrier name is " << name_ + "_barrier" << std::endl;
+
+                hpx::lcos::barrier
+                    barrier(name_ + "_barrier" , num_images_, image_id);
+                spmd_block block(num_images_, image_id, barrier);
+
+                int * dummy = nullptr;
+                hpx::util::invoke(
+                    reinterpret_cast<const F&>(*dummy),
+                    std::move(block),
+                    std::forward<Ts>(ts)...);
+
+                // Ensure that other images reaches that point
+                barrier.wait();
+            }
+        };
+    }
+
+    template <typename ExPolicy, typename F, typename ... Args>
+    void define_spmd_block(std::string && name, ExPolicy &&,
+        std::size_t images_per_locality, F && f, Args && ... args)
+    {
+        static_assert(
+            parallel::execution::is_execution_policy<ExPolicy>::value,
+            "parallel::execution::is_execution_policy<ExPolicy>::value");
+
+        using ftype = typename std::decay<F>::type;
+
+        using first_type =
+            typename hpx::parallel::v2::detail::extract_first_parameter<
+                        decltype(&ftype::operator())>::type;
+
+        using executor_type =
+            typename hpx::util::decay<ExPolicy>::type::executor_type;
+
+        static_assert(std::is_same<spmd_block, first_type>::value,
+            "define_spmd_block() needs a lambda that " \
+            "has at least a spmd_block as 1st argument");
+
+        std::size_t num_images
+            = hpx::get_num_localities(hpx::launch::sync) * images_per_locality;
+
+        // Force f to be initialized at compile-time
+        auto dummy = hpx::actions::lambda_to_action(f);
+        HPX_UNUSED(dummy);
+
+        auto act = hpx::actions::lambda_to_action(
+            []( std::string name,
+                std::size_t images_per_locality,
+                std::size_t num_images,
+                Args... args)
+            {
+                executor_type exec;
+                std::size_t offset = hpx::get_locality_id();
+                offset *= images_per_locality;
+
+                return
+                    hpx::parallel::executor_traits<
+                        executor_type
+                        >::bulk_execute(
```

Comment:
Maybe a good idea but for now I'm testing the synchronous version to make sure that everything is working fine. Got a deadlock when the `.wait()` `lcos::barrier` method is called.